lib/Dialect/AIE/Transforms/AIEAssignLockIDs.cpp

```diff
@@ -40,50 +39,69 @@ struct AIEAssignLockIDsPass : AIEAssignLockIDsBase<AIEAssignLockIDsPass> {
+    // assigned to a lock, and ops which have not.
+    struct TileLockOps {
+      DenseSet<int> assigned;
+      SmallVector<LockOp> unassigned;
+    };
+
+    DenseMap<TileOp, TileLockOps> tileToLocks;
 
-    // The second pass scans for locks with no lockIDs and assigns locks.
-    for (auto lock : device.getOps<LockOp>()) {
-      Operation *lock_tile = lock.getTile().getDefiningOp();
-      if (!lock.getLockID().has_value()) {
-        if (tileToLastID.find(lock_tile) == tileToLastID.end()) {
-          // If the tile operation corresponding to the lock does not exist in
-          // the data structure, initialize the lockID with 0 with an empty set.
-          tileToLastID[lock_tile].first = 0;
-        } else if (tileToLastID[lock_tile].first < 15) {
-          // If the tile operation of the lock exists, the potential lockID is
-          // checked with the set containing occupied lockIDs until a lockID
-          // that is free is found.
-          int potential_ID = tileToLastID[lock_tile].first;
-          while (true) {
-            if (tileToLastID[lock_tile].second.find(potential_ID) !=
-                tileToLastID[lock_tile].second.end())
-              potential_ID++;
-            else
-              break;
+    // Construct data structure storing locks by tile.
+    for (auto lockOp : device.getOps<LockOp>()) {
+
+      auto tileOp = lockOp.getTileOp();
+      bool isAssigned = lockOp.getLockID().has_value();
+
+      if (isAssigned) {
+        auto lockID = lockOp.getLockID().value();
+        auto iter = tileToLocks.find(tileOp);
+        if (iter == tileToLocks.end()) {
+          tileToLocks.insert({tileOp, {{lockID}, /* unassigned = */ {}}});
+        } else {
+          if (iter->second.assigned.find(lockID) !=
+              iter->second.assigned.end()) {
+            auto diag = lockOp->emitOpError("is assigned to the same lock (")
+                        << lockID << ") as another op.";
+            diag.attachNote(tileOp.getLoc())
+                << "tile has lock ops assigned to same lock.";
+            return signalPassFailure();
           }
-          tileToLastID[lock_tile].first = potential_ID;
+          iter->second.assigned.insert(lockID);
+        }
+      }
+
+      else {
+        auto iter = tileToLocks.find(tileOp);
+        if (iter == tileToLocks.end()) {
+          tileToLocks.insert({tileOp, {/* assigned = */ {}, {lockOp}}});
         } else {
-          lock->emitError() << "Exceeded the number of unique LockIDs";
-          return;
+          iter->second.unassigned.push_back(lockOp);
         }
+      }
+    }
 
-        // The lockID is assigned and is stored in the set.
-        lock->setAttr("lockID", rewriter.getI32IntegerAttr(
-                                    tileToLastID[lock_tile].first));
-        tileToLastID[lock_tile].second.insert(tileToLastID[lock_tile].first);
+    // IR mutation: assign locks to all unassigned lock ops.
+    for (auto &&[tileOp, locks] : tileToLocks) {
```

Comment:
I believe since `tileToLocks` is `DenseMap` the range-for will give you references even with a structured binding so you don't need the `&&` (not sure).

Reply:
I often use the universal reference `&&` as a just-do-the-right thing. Let the compiler deduce the reference type.  

You're right that DenseMap (and as far as I know all common map type) will give references, and it would require a weird custom map class for the `auto [tileOp, locks]` case to actually incur a copy of locks. 

Also, it seems like the universal reference approach is not common upstream (251 vs 31 in llvm/mlir/lib) so I'll change this case to what you suggest.